be/src/runtime/batch_write/txn_state_cache.h

```diff
@@ -0,0 +1,273 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "common/utils.h"
+#include "testutil/sync_point.h"
+#include "util/bthreads/bthread_shared_mutex.h"
+#include "util/dynamic_cache.h"
+#include "util/threadpool.h"
+#include "util/thrift_rpc_helper.h"
+
+namespace starrocks {
+
+class ThreadPoolToken;
+class TxnStateHandler;
+class TxnStateSubscriber;
+class TxnStateCache;
+using TxnStateDynamicCache = DynamicCache<int64_t, TxnStateHandler>;
+using TxnStateDynamicCachePtr = std::unique_ptr<TxnStateDynamicCache>;
+using TxnStateDynamicCacheEntry = TxnStateDynamicCache::Entry;
+
+struct TxnState {
+    TTransactionStatus::type txn_status{TTransactionStatus::PREPARE};
+    std::string reason;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const TxnState& txn_state) {
+    os << "txn_status: " << to_string(txn_state.txn_status) << ", reason: " << txn_state.reason;
+    return os;
+}
+
+// Handle the transaction state. It's the value of the DynamicCache entry
+// 1. transmit the txn state according to the new state pushed by FE or polled from FE
+// 2. notify txn state subscribers whether it reaches the finished state
+// 3. control the behaviour of txn state poll. The poll task starts to schedule when the
+//    first subscriber comes(see subscribe()), and continue to schedule (see poll_state())
+//    until the txn state reaches the finished state, and there is no subscriber
+class TxnStateHandler {
+public:
+    ~TxnStateHandler();
+
+    // update the txn state pushed by FE
+    void push_state(TTransactionStatus::type new_status, const std::string& reason);
+    // update the txn state polled by the cache. The returned value tell the caller to
+    // continue or stop polling according the current result and the current txn state
+    bool poll_state(const StatusOr<TxnState>& result);
+
+    // Add a subscriber for the finished state. Handler will set 'trigger_poll'
+    // to tell whether the caller should submit a txn state poll task
+    void subscribe(bool& trigger_poll);
```

Comment:
```suggestion
    [[nodiscard]] bool subscribe();
```

Reply:
I have thought to use the return value, but I want to make it more notable that subscribe can trigger a poll task. So I implement it with a named parameter